### DISCUSSION TITLE
Using the Newton solver for k-epsilon turbulence models

### DISCUSSION THREAD

**Discussion by charlieowen3065** [idaholab/moose#29236] [Q&A Modules: Navier-Stokes] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A Navier Stokes is the most appropriate category for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (avoid screenshots if possible, triple back quotes before/after pasted text, etc)

### Question

Hello!

I am currently attempting to design a RANS model using the k-epsilon formulation you all have implemented recently, and had a question on the matter.

For my specific purposes, it would be useful to use the Newton-solver to simulate my experiment (I am attempting to do some ROM analysis, and algorithms I have written in the past are based on the Newton update). For non-RANS models (just using simple mass, momentum, and heat transfers), the Newton algorithm seemed to be working, but once I added in TKE and TKED variables and kernels, it gave me the error
```
For iteration type 'nonlinear': solution state 1 was requested in nl0 but only up to state 0 is available.
```
And it looks like the error is being given by the `INSFVTKESourceSink` object. Are Newton capabilities not implemented for the k-epsilon model yet? Or is there a work-around that I have not found?

I've seen the SIMPLE algorithm, and I could probably work with it, but it would be nice to work with what I'm familiar with.

Any ideas on where to go with this?

Thanks!

Charlie

**Comment by GiudGiud** [comment]
> And it looks like the error is being given by the INSFVTKESourceSink object. Are Newton capabilities not implemented for the k-epsilon model yet?

They are enabled, though we do not leverage them anywhere so you would be a first-adopter

>  Or is there a work-around that I have not found?

Did you set the parameters in the problem to request the previous solution? Did you follow the tests for TKE + Newton ?

**Comment by charlieowen3065** [comment]
I did not manually set the parameters in the problem to request the previous solution - I was not aware of that functionality, but I will look into that!

For the tests of TKE + Newton, they all use the `Physics` block. To build my model, I wanted to start at the basics (I started with just steady-state mass-momentum) and have been building my model up from there, and recently reached the point where RANS was being necessary, and in that, I have not been using the physics block. Does this block work similar to the `Modules` block, where it is a short-hand for a larger set of input blocks (such that you could completely rebuild it with just blocks in the input file), or is it made deeper in the code?

I was basing my Turbulence models on the `channel_ERCOFTAC.i` channel-test, and was trying to rework that to use Newton updates.

I am mostly trying to learn how to properly use the Navier-Stokes module right now, so I was avoiding using the larger `Module`/`Physics` blocks to have a better handle of what is happening under the hood, and to try and do minor parametric studies to see how much various inputs affect response.

**Reply by GiudGiud** [reply, marked as answer]
> For the tests of TKE + Newton, they all use the Physics block.

once https://github.com/idaholab/moose/pull/29177 is merged you ll be able to dump the Physics syntax to regular kernel syntax. It might already work in your case. Try setting (just once) Problem/type=DumpObjectsProblem

Regardless, you can see the parameters of the Problem and executioner block and try to use those

> Does this block work similar to the Modules block, where it is a short-hand for a larger set of input blocks (such that you could completely rebuild it with just blocks in the input file), or is it made deeper in the code?

No it's exactly the same as the Modules block, except that there's tons of additional helper and checks that we did not have back when using the Modules block

**Reply by charlieowen3065** [reply]
Okay! the dump wasn't quite working, but it was a mpicxx error, which I haven't set up yet, so that may be on my end. I'll look into that.

I see the source file, so I'll just parse through that in the meantime - since it does work though, I'm all good! 

Thank you!

Charlie

**Comment by ysamolada** [comment]
I'm also encountering the following error when running a simulation using the finite volume (FV) formulation in MOOSE:

*** ERROR ***
For iteration type 'nonlinear': solution state 1 was requested in nl0 but only up to state 0 is available.

Context:
This error occurs at the start of the simulation, before any time stepping progresses. My model is a 2D fluid-thermal simulation using the INSFV module with buoyancy, turbulence (k–ε), and heat source terms.

The issue arises when I try implementing the k and epsilon FVKernels, and specifically the sink. When I comment out the sink, the model runs but the values for viscosity are obviously off. 


Any tips on debugging this kind of issue or properly ensuring the required solution states are available would be appreciated.

Thanks in advance!

**Reply by GiudGiud** [reply]
hello

You can save the previous solution with 
```
[Problem]
  previous_nl_solution_required = true
[]
```

**Reply by charlieowen3065** [reply]
I personally never figured out what the exact error was that was causing this, but what I was doing (and what it sounds like you are doing) was trying to build up the entire physics of the system in the `Kernels` block.

I know theoretically this can definitely be done, but it seemed like it might require some extra bits or have really specific ways to build it. What I ended up doing was using the `Physics` block to define the fluid flow, heat transfer, and turbulence. These blocks were designed to give a lot of freedom while adding in necessary parts or specifics that could be missed.

I would suggest trying that out (there are a lot of options in the three `Physics` sub-blocks, so there is still plenty of freedom), and to debug the precise issue, you could go through the `Physics` source code for turbulent physics. Within this code, it has function calls for each type of `Kernel`, 'BC', parameter, etc in it, and you should be able to go through that and build it up from there.

I never ended up doing that as the `Physics` block had everything I needed for my work, but I did skim through it and it did make sense.

Hope this helps!

Charlie
